packages/medusa/src/api/utils/middlewares/products/set-tax-context.ts

```diff
@@ -0,0 +1,75 @@
+import { TaxCalculationContext } from "@medusajs/types"
+import { NextFunction } from "express"
+import {
+  AuthenticatedMedusaRequest,
+  MedusaRequest,
+} from "../../../../types/routing"
+import { refetchEntity } from "../../refetch-entity"
+import { MedusaError } from "@medusajs/utils"
+import { RequestWithContext } from "../../../store/products/helpers"
+
+export function setTaxContext() {
+  return async (req: AuthenticatedMedusaRequest, _, next: NextFunction) => {
+    const withCalculatedPrice = req.remoteQueryConfig.fields.some((field) =>
+      field.startsWith("variants.calculated_price")
+    )
+    if (!withCalculatedPrice) {
+      return next()
+    }
+
+    try {
+      const inclusivity = await getTaxInclusivityInfo(req)
+      if (!inclusivity || !inclusivity.automaticTaxes) {
+        return next()
+      }
+
+      const taxLinesContext = await getTaxLinesContext(req)
+
+      // TODO: Allow passing a context typings param to AuthenticatedMedusaRequest
+      ;(req as unknown as RequestWithContext<any>).taxContext = {
+        taxLineContext: taxLinesContext,
+        taxInclusivityContext: inclusivity,
+      }
+      return next()
+    } catch (e) {
+      next(e)
+    }
+  }
+}
+
+const getTaxInclusivityInfo = async (req: MedusaRequest) => {
+  const region = await refetchEntity(
+    "region",
+    req.filterableFields.region_id as string,
+    req.scope,
+    ["automatic_taxes"]
+  )
+
+  if (!region) {
+    throw new MedusaError(
+      MedusaError.Types.INVALID_DATA,
+      `Region with id ${req.filterableFields.region_id} not found when populating the tax context`
+    )
+  }
+
+  // TODO: Add is_tax_inclusive here from pricing module
+  return {
+    isTaxInclusive: true,
```

Comment:
question: Just for clarity, how would this flag get populated in a future iteration?

Reply:
@olivermrbl as I posted in the tax inclusive channel, we will store this as part of the pricing module (to begin with, set with either a currency or region rule), and we will just fetch it here before returning. The rest of the logic should remain unchanged (this will happen in a separate PR)

Reply:
As a separate data model in the pricing module right?

Reply:
Yes exactly, that is also rule-based (at least that was the last thing we discussed with Seb)

Reply:
Makes sense